edition = "2023";

package pbrt_proto.v3;

message Accelerator {
  message Bvh {
    enum SplitMethod {
      option features.enum_type = CLOSED;
      SAH = 0;
      MIDDLE = 1;
      EQUAL = 2;
      HLBVH = 3;
    }

    int32 maxnodeprims = 1 [default = 4];
    SplitMethod splitmethod = 2;
  }

  message KdTree {
    int32 intersectcost = 1 [default = 80];
    int32 traversalcost = 2 [default = 1];
    double emptybonus = 3 [default = 0.2];
    int32 maxprims = 4 [default = 1];
    int32 maxdepth = 5 [default = -1];
  }

  oneof accelerator_type {
    Bvh bvh = 1;
    KdTree kdtree = 2;
  }
}

message ActiveTransform {
  enum ActiveTransformation {
    option features.enum_type = CLOSED;
    ALL = 1;
    START_TIME = 2;
    END_TIME = 3;
  };

  ActiveTransformation active = 1 [features.field_presence = LEGACY_REQUIRED];
}

message ConcatTransform {
  double m00 = 1 [features.field_presence = LEGACY_REQUIRED];
  double m01 = 2 [features.field_presence = LEGACY_REQUIRED];
  double m02 = 3 [features.field_presence = LEGACY_REQUIRED];
  double m03 = 4 [features.field_presence = LEGACY_REQUIRED];
  double m10 = 5 [features.field_presence = LEGACY_REQUIRED];
  double m11 = 6 [features.field_presence = LEGACY_REQUIRED];
  double m12 = 7 [features.field_presence = LEGACY_REQUIRED];
  double m13 = 8 [features.field_presence = LEGACY_REQUIRED];
  double m20 = 9 [features.field_presence = LEGACY_REQUIRED];
  double m21 = 10 [features.field_presence = LEGACY_REQUIRED];
  double m22 = 11 [features.field_presence = LEGACY_REQUIRED];
  double m23 = 12 [features.field_presence = LEGACY_REQUIRED];
  double m30 = 13 [features.field_presence = LEGACY_REQUIRED];
  double m31 = 14 [features.field_presence = LEGACY_REQUIRED];
  double m32 = 15 [features.field_presence = LEGACY_REQUIRED];
  double m33 = 16 [features.field_presence = LEGACY_REQUIRED];
}

message CoordinateSystem {
  string name = 1 [features.field_presence = LEGACY_REQUIRED];
}

message CoordSysTransform {
  string name = 1 [features.field_presence = LEGACY_REQUIRED];
}

message Identity {
  // Empty Message
}

message Include {
  string path = 1 [features.field_presence = LEGACY_REQUIRED];
}

message LookAt {
  double eye_x = 1 [features.field_presence = LEGACY_REQUIRED];
  double eye_y = 2 [features.field_presence = LEGACY_REQUIRED];
  double eye_z = 3 [features.field_presence = LEGACY_REQUIRED];
  double look_x = 4 [features.field_presence = LEGACY_REQUIRED];
  double look_y = 5 [features.field_presence = LEGACY_REQUIRED];
  double look_z = 6 [features.field_presence = LEGACY_REQUIRED];
  double up_x = 7 [features.field_presence = LEGACY_REQUIRED];
  double up_y = 8 [features.field_presence = LEGACY_REQUIRED];
  double up_z = 9 [features.field_presence = LEGACY_REQUIRED];
}

message ReverseOrientation {
  // Empty Message
}

message Rotate {
  double angle = 1 [features.field_presence = LEGACY_REQUIRED];
  double x = 2 [features.field_presence = LEGACY_REQUIRED];
  double y = 3 [features.field_presence = LEGACY_REQUIRED];
  double z = 4 [features.field_presence = LEGACY_REQUIRED];
}

message Scale {
  double x = 1 [features.field_presence = LEGACY_REQUIRED];
  double y = 2 [features.field_presence = LEGACY_REQUIRED];
  double z = 3 [features.field_presence = LEGACY_REQUIRED];
}

message Transform {
  double m00 = 1 [features.field_presence = LEGACY_REQUIRED];
  double m01 = 2 [features.field_presence = LEGACY_REQUIRED];
  double m02 = 3 [features.field_presence = LEGACY_REQUIRED];
  double m03 = 4 [features.field_presence = LEGACY_REQUIRED];
  double m10 = 5 [features.field_presence = LEGACY_REQUIRED];
  double m11 = 6 [features.field_presence = LEGACY_REQUIRED];
  double m12 = 7 [features.field_presence = LEGACY_REQUIRED];
  double m13 = 8 [features.field_presence = LEGACY_REQUIRED];
  double m20 = 9 [features.field_presence = LEGACY_REQUIRED];
  double m21 = 10 [features.field_presence = LEGACY_REQUIRED];
  double m22 = 11 [features.field_presence = LEGACY_REQUIRED];
  double m23 = 12 [features.field_presence = LEGACY_REQUIRED];
  double m30 = 13 [features.field_presence = LEGACY_REQUIRED];
  double m31 = 14 [features.field_presence = LEGACY_REQUIRED];
  double m32 = 15 [features.field_presence = LEGACY_REQUIRED];
  double m33 = 16 [features.field_presence = LEGACY_REQUIRED];
}

message TransformTimes {
  double start_time = 1 [features.field_presence = LEGACY_REQUIRED];
  double end_time = 2 [features.field_presence = LEGACY_REQUIRED];
}

message Translate {
  double x = 1 [features.field_presence = LEGACY_REQUIRED];
  double y = 2 [features.field_presence = LEGACY_REQUIRED];
  double z = 3 [features.field_presence = LEGACY_REQUIRED];
}

message WorldBegin {
  // Empty Message
}

message WorldEnd {
  // Empty Message
}

message Directive {
  oneof directive_type {
    Accelerator accelerator = 1;
    ActiveTransform active_transform = 2;
    ConcatTransform concat_transform = 3;
    CoordinateSystem coordinate_system = 4;
    CoordSysTransform coord_sys_transform = 5;
    Identity identity = 6;
    Include include = 7;
    LookAt look_at = 8;
    ReverseOrientation reverse_orientation = 9;
    Rotate rotate = 10;
    Scale scale = 11;
    Transform transform = 12;
    TransformTimes transform_times = 13;
    Translate translate = 14;
    WorldBegin world_begin = 15;
    WorldEnd world_end = 16;
  }
}

message PbrtProto {
  repeated Directive directives = 1;
}
